### PR TITLE
[Enhancement] Optimize memory tracker

### PR DESCRIPTION
**Pull request by gengjun-git**
## Why I'm doing:
Currently, memory statistics consume a lot of performance because they are all calculated each time.

## What I'm doing:
Get some samples to estimate the manager's memory, even though it is inaccurate, but it's helpful to solve memory leak bugs.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

## Documentation PRs only:

If you are submitting a PR that adds or changes English documentation and have not
included Chinese documentation, then you can check the box to request GPT to translate the
English doc to Chinese. Please ensure to uncheck the **Do not translate** box if translation is needed.
The workflow will generate a new PR with the Chinese translation after this PR is merged.

- [ ] Yes, translate English markdown files with GPT
- [x] Do not translate

### CODE REVIEW

**Review by Astralidea — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java [hunk 1]
```diff
@@ -14,18 +14,28 @@
...
+        }
 
-    default Map<String, Long> estimateCount() {
-        return new HashMap<>();
+        return totalBytes;
     }
+
+    Map<String, Long> estimateCount();
+
+    List<Pair<List<Object>, Long>> getSamples();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java:40]: plz add some comments for this api

File: fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java [hunk 2]
```diff
@@ -14,18 +14,28 @@
...
 
 public interface MemoryTrackable {
-
     default long estimateSize() {
-        return SizeEstimator.estimate(this);
-    }
+        List<Pair<List<Object>, Long>> samples = getSamples();
+        long totalBytes = 0;
+        for (Pair<List<Object>, Long> pair : samples) {
+            List<Object> sample = pair.first;
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java:28]:
```suggestion
            List<Object> sampleObjects = pair.first;
```

File: fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java [hunk 3]
```diff
@@ -14,18 +14,28 @@
...
+            long size = pair.second;
+            if (sample.size() > 0) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java:30]: Precondition.check?

File: fe/fe-core/src/main/java/com/starrocks/memory/MemoryUsageTracker.java [hunk 4]
```diff
@@ -91,11 +95,16 @@ public static void registerMemoryTracker(String moduleName, MemoryTrackable obje
     }
 
     public static void trackMemory() {
-        trackMemory(REFERENCE);
-        trackMemory(ImmutableMap.of("Connector", GlobalStateMgr.getCurrentState().getConnectorMgr().getMemTrackers()));
+        long totalTracked = trackMemory(REFERENCE);
+        totalTracked += trackMemory(ImmutableMap.of("Connector",
+                GlobalStateMgr.getCurrentState().getConnectorMgr().getMemTrackers()));
+
+        LOG.info("total tracked memory: {}, heap used memory: {}", new ByteSizeValue(totalTracked),
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/memory/MemoryUsageTracker.java:102]: Add another 2 metrics?
1. RSS of the java process
2. Size of off-heap memory, like direct buffer pool, etc.

File: fe/fe-core/src/main/java/com/starrocks/memory/InternalCatalogMemoryTracker.java [hunk 5]
```diff
@@ -1,64 +0,0 @@
...
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-
-public class InternalCatalogMemoryTracker implements MemoryTrackable {
-
-    @Override
-    public long estimateSize() {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/memory/InternalCatalogMemoryTracker.java:33]: we still need to calculate the size of all the partitions except the replicas in it

File: fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java (hunk 3 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/memory/MemoryTrackable.java:30]: No, empty is normal when there is no objects in that manager.

**Review by nshangyiming — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Review by meegoo — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java [hunk 6]
```diff
@@ -65,6 +65,8 @@ public class CachingIcebergCatalog implements IcebergCatalog {
     private static final Logger LOG = LogManager.getLogger(CachingIcebergCatalog.class);
     public static final long NEVER_CACHE = 0;
     public static final long DEFAULT_CACHE_NUM = 100000;
+    private static final int MEMORY_META_SAMPLES = 10;
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/CachingIcebergCatalog.java:68]: why not use a config?

**Review by nshangyiming — APPROVED** [review]

**Comment by gengjun-git** [comment]
@mergifyio rebase

**Review by meegoo — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Pull request merged**